jerry-core/ecma/base/ecma-helpers.c

```diff
@@ -842,34 +875,61 @@ ecma_free_property (ecma_object_t *object_p, /**< object the property belongs to
  * Warning: specified property must be owned by specified object.
  */
 void
-ecma_delete_property (ecma_object_t *obj_p, /**< object */
+ecma_delete_property (ecma_object_t *object_p, /**< object */
                       ecma_property_t *prop_p) /**< property */
 {
-  for (ecma_property_t *cur_prop_p = ecma_get_property_list (obj_p), *prev_prop_p = NULL, *next_prop_p;
-       cur_prop_p != NULL;
-       prev_prop_p = cur_prop_p, cur_prop_p = next_prop_p)
+  ecma_property_header_t *cur_prop_p = ecma_get_property_list (object_p);
+  ecma_property_header_t *prev_prop_p = NULL;
+
+  while (true)
   {
-    next_prop_p = ECMA_GET_POINTER (ecma_property_t,
-                                    cur_prop_p->next_property_p);
+    JERRY_ASSERT (cur_prop_p != NULL);
+    JERRY_ASSERT (ECMA_PROPERTY_IS_PROPERTY_PAIR (cur_prop_p));
 
-    if (cur_prop_p == prop_p)
-    {
-      ecma_free_property (obj_p, prop_p);
+    ecma_property_pair_t *prop_pair_p = (ecma_property_pair_t *) cur_prop_p;
 
-      if (prev_prop_p == NULL)
-      {
-        ecma_set_property_list (obj_p, next_prop_p);
-      }
-      else
+    for (int i = 0; i < 2; i++)
+    {
+      if ((cur_prop_p->types + i) == prop_p)
       {
-        ECMA_SET_POINTER (prev_prop_p->next_property_p, next_prop_p);
-      }
+        ecma_string_t *name_p = ECMA_GET_POINTER (ecma_string_t, prop_pair_p->names_cp[i]);
+
+        ecma_free_property (object_p, name_p, cur_prop_p->types + i);
+
+        prop_pair_p->names_cp[i] = ECMA_NULL_POINTER;
+
+        if (name_p != NULL)
+        {
+          ecma_deref_ecma_string (name_p);
+        }
+
+        if (cur_prop_p->types[1 - i].type_and_flags != ECMA_PROPERTY_TYPE_DELETED)
+        {
+          /* The other property is still valid. */
+          return;
+        }
 
-      return;
+        JERRY_ASSERT (cur_prop_p->types[0].type_and_flags == ECMA_PROPERTY_TYPE_DELETED
+                      && cur_prop_p->types[1].type_and_flags == ECMA_PROPERTY_TYPE_DELETED);
```

Comment:
This assert checks types_and_flags twice. It's enough to assert on `cur_prop_p->types[i].type_and_flags == ECMA_PROPERTY_TYPE_DELETED` since we are in the loop.